support clickhouse as mysql federated server

MaxFedotov opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (up to few sentences, not needed for non-significant PRs):
Support adding Clickhouse as MySQL federated server (see https://dev.mysql.com/doc/refman/5.7/en/federated-create-server.html)

Commit:
support clickhouse as mysql federated server

alexey-milovidov commented:
That's very interesting feature!

@yurriy will help with review.

yurriy commented on dbms/programs/server/MySQLHandler.cpp:
```diff
@@ -266,29 +267,56 @@ void MySQLHandler::comPing()
+        if (0 == strncasecmp("SHOW TABLE STATUS LIKE", query.c_str(), 22))
+        {
+            should_replace = true;
+            replacement_query = boost::replace_all_copy(query, "SHOW TABLE STATUS LIKE ", "SELECT name AS Name, engine AS Engine, '10' AS Version, 'Dynamic' AS Row_format, 0 AS Rows, 0 AS Avg_row_length, 0 AS Data_length, 0 AS Max_data_length, 0 AS Index_length, 0 AS Data_free, 'NULL' AS Auto_increment, metadata_modification_time AS Create_time, metadata_modification_time AS Update_time, metadata ...
```
I would split the third argument into several lines.

MaxFedotov replied:
fixed

yurriy reviewed (changes requested):
Nice feature. Thanks. Could you please add a test?
Tests of MySQL protocol are [here](https://github.com/ClickHouse/ClickHouse/blob/master/dbms/tests/integration/test_mysql_protocol/test.py). MySQL server will be started in mysql_client docker container if [this line](https://github.com/ClickHouse/ClickHouse/blob/master/dbms/tests/integration/test_mysql_protocol/clients/mysql/docker_compose.yml#L6) is removed. [dbms/tests/integration/README.md](https://github.com/ClickHouse/ClickHouse/blob/master/dbms/tests/integration/README.md).

Commit:
support clickhouse as mysql federated server. apply code review notes

MaxFedotov commented on dbms/programs/server/MySQLHandler.cpp:
```diff
@@ -266,29 +267,56 @@ void MySQLHandler::comPing()
 
-    const String query("select ''");
-    ReadBufferFromString empty_select(query);
+        String replacement_query = ("select ''");
```
fixed

MaxFedotov commented:
Hi, @yurriy. Sorry, was on vacation and only now was able to get back to this commit. I've fixed all issues, that you highlighted and added tests.

Thanks, 
Max

Commit:
support clickhouse as mysql federated server. apply code review notes. fix typo

yurriy commented on dbms/programs/server/MySQLHandler.cpp:
```diff
@@ -266,29 +267,78 @@ void MySQLHandler::comPing()
+            replacement_query = boost::replace_all_copy(query, 
+                                                        "SHOW TABLE STATUS LIKE ", 
+                                                        "SELECT \
+                                                            name AS Name, \
```
Quotes are needed on each line to avoid including unnecessary spaces in the query

yurriy replied:
Let's move the replacement query to a constant variable at the end of a file, to keep comQuery function short.

MaxFedotov replied:
done

yurriy commented on dbms/tests/integration/test_mysql_protocol/clients/mysql/docker_compose.yml:
```diff
@@ -1,6 +1,8 @@
+      restart: always
+      environment:
+          MYSQL_ALLOW_EMPTY_PASSWORD: 1
+      command: --federated --socket /var/run/mysqld/mysqld.sock
```
Indent differs from an indent on line 3
Newline at the end of file is missing

MaxFedotov replied:
done

yurriy commented on dbms/programs/server/MySQLHandler.cpp:
```diff
@@ -266,29 +267,78 @@ void MySQLHandler::comPing()
+        (0 == strncasecmp("SET AUTOCOMMIT", query.c_str(), 14))
+        ||
+        (0 == strncasecmp("SET SESSION TRANSACTION ISOLATION LEVEL", query.c_str(), 39))
+    ) 
```
Should be moved to a static function isFederatedServerSetupCommand, to keep comQuery function short.

MaxFedotov replied:
done

yurriy commented on dbms/programs/server/MySQLHandler.cpp:
```diff
@@ -266,29 +267,78 @@ void MySQLHandler::comPing()
-        with_output = true;
-    };
+void MySQLHandler::comQuery(ReadBuffer &payload) {
+    std::string query = std::string(payload.position(), payload.buffer().end());
```
std::string should be replaced by String for consistency. It's the same class.

MaxFedotov replied:
done

yurriy reviewed (changes requested):
Note that there were few style check errors, they are listed in "Details" link below.

Commit:
support clickhouse as mysql federated server. apply code review notes p2

Commit:
support clickhouse as mysql federated server. fix code style

yurriy commented on dbms/programs/server/MySQLHandler.cpp:
```diff
@@ -266,29 +267,60 @@ void MySQLHandler::comPing()
 }
 
-void MySQLHandler::comQuery(ReadBuffer & payload)
+static bool isFederatedServerSetupCommand(String query)
```
const String &

MaxFedotov replied:
done

yurriy commented on dbms/programs/server/MySQLHandler.cpp:
```diff
@@ -266,29 +267,60 @@ void MySQLHandler::comPing()
-    std::function<void(const String &)> set_content_type = [&with_output](const String &) -> void {
-        with_output = true;
-    };
+    if ((0 == strncasecmp("SET NAMES", query.c_str(), 9)) || (0 == strncasecmp("SET character_set_results", query.c_str(), 25))
```
`if` is redundant, it can be just a `return` statement.
`return (0 == strncasecmp("SET NAMES", query.c_str(), 9)) || ...`

MaxFedotov replied:
done

yurriy commented on dbms/programs/server/MySQLHandler.cpp:
```diff
@@ -334,4 +366,26 @@ void MySQLHandlerSSL::finishHandshakeSSL(size_t packet_size, char * buf, size_t
 
 #endif
 
+const String show_table_status_replacement_query("SELECT "
```
Extra space after SELECT

yurriy replied:
Forward declaration is required in MySQLHandler.h

MaxFedotov replied:
done

yurriy commented on dbms/programs/server/MySQLHandler.cpp:
```diff
@@ -266,29 +267,60 @@ void MySQLHandler::comPing()
+        if (0 == strncasecmp("SHOW TABLE STATUS LIKE", query.c_str(), 22))
+        {
+            should_replace = true;
+            replacement_query = boost::replace_all_copy(query,
```
Can be placed on a single line now

MaxFedotov replied:
done

Commit:
support clickhouse as mysql federated server. apply code review notes p3

MaxFedotov commented:
Hi @yurriy, sorry for stupid mistakes - hadn't written on C++ since university :) I've applied all code review notes, can you please take a look?

Thank,
Max

yurriy commented on dbms/programs/server/MySQLHandler.h:
```diff
@@ -77,4 +79,26 @@ class MySQLHandlerSSL : public MySQLHandler
 };
 #endif
 
+const String show_table_status_replacement_query("SELECT"
```
It was fine to have a definition in .cpp, so that the header file is short.
But forward declaration was required to use this variable:
`const String show_table_status_replacement_query;`
since the definition was at the end of file.

But it would be even better to make this variable static and to put both declaration and definition to MySQLHandler.cpp. And the same for isFederatedServerSetupCommand:
```
static bool isFederatedServerSetupCommand(const String &query);
static const String show_table_status_replacement_query;

comQuery()
{
}

...

static bool isFederatedServerSetupCommand(const String &query)
{
}

static const String show_table_status_replacement_query("SELECT"...
```
Motivation in having a separate declaration and definition: handler code will still be compact and will be located in the beginning of file.

MaxFedotov replied:
@yurriy if I try to do as you suggested - move both declaration and definition to MySQLHandler.cpp (see https://gist.github.com/MaxFedotov/8ae6da39cd5c45c3d3fcaf358181298a), i've got the following error during compilation:
`
../dbms/programs/server/MySQLHandler.cpp:366:21: error: redefinition of 'const String DB::show_table_status_replacement_query'
  366 | static const String show_table_status_replacement_query("SELECT"
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../dbms/programs/server/MySQLHandler.cpp:271:21: note: 'const String DB::show_table_status_replacement_query' previously declared here
  271 | static const String show_table_status_replacement_query;
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
`

yurriy replied:
Sorry, it doesn't work for variables. Should work if show_table_status_replacement_query will be a static private member of MySQLHandler.
```
// In MySQLHandler.h
class MySQLHandler
{
...
private:
    static const String MySQLHandler::show_table_status_replacement_query;
}

// At the end of MySQLHandler.cpp
const String MySQLHandler::show_table_status_replacement_query("SELECT"
```

MaxFedotov replied:
seems like it should be 
`
private:
    static const String show_table_status_replacement_query;
`
, or during build following error occurs:
`../dbms/programs/server/MySQLHandler.h:63:25: error: extra qualification 'DB::MySQLHandler::' on member 'show_table_status_replacement_query' [-fpermissive]
   63 |     static const String MySQLHandler::show_table_status_replacement_query;
      |                         ^~~~~~~~~~~~
ninja: build stopped: subcommand failed.`

yurriy commented on dbms/programs/server/MySQLHandler.cpp:
```diff
@@ -267,15 +267,11 @@ void MySQLHandler::comPing()
+static bool isFederatedServerSetupCommand(const String &query)
 {
-    if ((0 == strncasecmp("SET NAMES", query.c_str(), 9)) || (0 == strncasecmp("SET character_set_results", query.c_str(), 25))
+    return ((0 == strncasecmp("SET NAMES", query.c_str(), 9)) || (0 == strncasecmp("SET character_set_results", query.c_str(), 25))
```
Brackets (around each condition and around the whole expression) are redundant

MaxFedotov replied:
fixed

yurriy commented on dbms/programs/server/MySQLHandler.h:
```diff
@@ -12,6 +12,8 @@
 namespace DB
 {
 
+
```
Too many empty lines (code style limit is 2 empty lines, but one line seems to be enough here)

MaxFedotov replied:
fixed

yurriy reviewed (changes requested):
> Hi @yurriy, sorry for stupid mistakes - hadn't written on C++ since university :) I've applied all code review notes, can you please take a look?
> 
> Thank,
> Max

No worries, you're welcome.

Commit:
support clickhouse as mysql federated server. apply code review notes p4

alexey-milovidov committed:
Update MySQLHandler.cpp

alexey-milovidov commented on dbms/programs/server/MySQLHandler.cpp:
```diff
@@ -266,29 +267,49 @@ void MySQLHandler::comPing()
+        if (0 == strncasecmp("SHOW TABLE STATUS LIKE", query.c_str(), 22))
+        {
+            should_replace = true;
+            replacement_query = boost::replace_all_copy(query, "SHOW TABLE STATUS LIKE ", show_table_status_replacement_query);
```
Here is LIKE. But in replacement query it is just a string equality.

alexey-milovidov replied:
Fixed.

alexey-milovidov committed:
Update MySQLHandler.cpp
